README.md
# `fastQ_brew v.2.0`

[![GitHub license](https://img.shields.io/badge/license-GPL_2.0-orange.svg)](https://raw.githubusercontent.com/dohalloran/fastQ_brew/master/LICENSE)
[![GitHub issues](https://img.shields.io/github/issues/dohalloran/fastQ_brew.svg)](https://github.com/dohalloran/fastQ_brew/issues)

- [x] `Pre-processing of FASTQ reads`
- [x] `Check that files were demultiplexed correctly`
- [x] `Filter reads by length `
- [x] `Filter reads by quality`
- [x] `Trim reads`
- [x] `Removes standard Truseq adapters`
- [x] `Performs various file conversions` 

![fastQ_brew LOGO](https://cloud.githubusercontent.com/assets/8477977/22077145/f29a177e-dd80-11e6-86a6-a211e8e1e103.jpg)

## Installation
1. Download and extract the fastQ_brew.zip file  
`tar -xzvf fastQ_brew.zip` 
or 
`git clone https://github.com/dohalloran/fastQ_brew.git`
2. The extracted dir will be called fastQ_brew  
```cmd  
cd fastQ_brew   
perl Makefile.PL  
make  
make test  
make install  
```

## Usage 
### To run:  
```perl
use fastQ_brew;
use Moose;
use Modern::Perl;
use autodie;

my $app = fastQ_brew->new_with_options();
$app->run_fastQ_brew(); 
#see below for command flags
``` 
## Command Line Arguments
### Filtering Options
 ```perl   
#set the max probability that a fastQ [1] read will contain errors: suggested p<=50% (must be 1-100)
        --qf 50
#filter by read length - reads below this length will be removed       
        --lf 35
#remove x bases from left end of every read 
        --trim_l 5
#remove x bases from right end of every read
        --trim_r 3
#remove standard truseq adapters (permits 1 mismatch) from both ends (very slow!)
        --truseq
```

### File Conversions and de-multiplex check
 ```perl   
#check that 2 FASTQ files were demultiplexed correctly 
#fastQ_brew outputs the barcodes for each file and compares (union and intersection) between two files 
        --plex
        -i <input_file1>
        -x <input_file2>
#convert FASTQ file to FASTA format file
        --fasta
#convert the DNA for each read to RNA 
        --dna_rna
#reverse complement the FASTQ reads 
        --rev_comp
```

### Odds and Ends
 ```perl   
#input FASTQ file (required) 
        --i <input_file>
#output FASTQ file (by default called filtered.fq) 
        --o <output_file>
#library type i.e. sanger (default) or illumina 
        --lib sanger
#print flag options to stdout
        --help  
```

### References
1. Cock PJ, Fields CJ, Goto N, Heuer ML, Rice PM. The Sanger FASTQ file format for sequences with quality scores, and the Solexa/Illumina FASTQ variants. Nucleic Acids Res. 2010;38(6):1767–71

2. Ewing B, Hillier L, Wendl MC, Green P. Base-calling of automated sequencer traces using phred. I. Accuracy assessment. Genome Res. 1998;8(3):175–85.

3. Ewing B, Green P. Base-calling of automated sequencer traces using phred. II. Error probabilities. Genome Res. 1998;8(3):186–94.

## Contributing
All contributions are welcome.

## Support
If you have any problem or suggestion please open an issue [here](https://github.com/dohalloran/fastQ_brew/issues).

## License 
GNU GENERAL PUBLIC LICENSE







Changes
Revision history for fastQ_brew

1.0     01/16/2017 
        First version, released on an unsuspecting world.

1.0.2   05/10/2017 
        Added new features to chain options.
        
1.0.3   06/02/2017 
        Added Getopt::Long features.
        
1.0.4   07/17/2017 
        Added File::Copy features.

1.0.5   09/02/2017 
        Consolidated subroutines to increase speed.

2.0.    09/05/2017 
        Modernized with Moose and made faster.




Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'fastQ_brew',
    AUTHOR           => q{Damien OHalloran <damienoh@gwu.edu>},
    VERSION_FROM     => 'lib/fastQ_brew.pm',
    ABSTRACT_FROM    => 'lib/fastQ_brew.pm',
    LICENSE          => 'gpl_2',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        'Sys::Hostname'           => 0,
        'File::Copy'              => 0,
        'Term::ANSIColor'         => 0,
        'Moose'                   => 2.0,
        'Modern::Perl'            => 1.0,
        'Text::LevenshteinXS'     => 0,
        'List::MoreUtils'         => 0,
        'Win32::Console::ANSI'    => 0,
        'Statistics::Basic'       => 0,
        'List::Compare'           => 0
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'fastQ_brew-*' },
);


lib/brew_driver.pl
#!/usr/bin/perl

use Moose;
use Modern::Perl;
use autodie;
use fastQ_brew;

my $app = fastQ_brew->new_with_options();
$app->run_fastQ_brew();

=head1 OPTIONS

  --i, input file (required)
  --x, input file2, only to check demultiplexing 
  --plex, checks that two FASTQ files were demultiplexed correctly
  --o, output file (required)    
  --lib, library type  (default is sanger)    
  --qf, filter by phred (suggested default=20, min Q score=8 by default and prob = 0.5)
  --lf, filter by read length (suggested default=25)
  --truseq, removes truseq adapters from reads
  --trim_l, trim reads starting at left end
  --trim_r, trim reads starting at left end   
  --fasta, convert to fastA format 
  --dna_rna, convert reads to RNA
  --rev_comp, reverse complement reads
  --help, Print this help
  
=cut


lib/fastQ_brew_Utilities.pm
#
# Utility module for fastQ_brew
#
# Please direct questions and support issues to <https://github.com/dohalloran/fastQ_brew/issues>
#
# Author: Damien O'Halloran, The George Washington University, 2017
#
# GNU GENERAL PUBLIC LICENSE
#
# POD documentation before the code

=head1 NAME

fastQ_brew_Utilities - utilities for fastQ_brew

=head2 SYNOPSIS

  use Moose;
  use Modern::Perl;
  use base 'Exporter';
  use List::Util qw(sum);
  use autodie;

=head2 DESCRIPTION

This package provides Utility support to fastQ_brew

=head2 Support

All contributions are welcome

=head2 Reporting Bugs

Report bugs to the fastQ_brew bug tracking system to help keep track
of the bugs and their resolution. Bug reports can be submitted via the
web:
  https://github.com/dohalloran/fastQ_brew/issues

=head1 APPENDIX

The rest of the documentation details each of the object methods.

=cut

package fastQ_brew_Utilities;

use Moose;
use Modern::Perl;
use base 'Exporter';
use List::Util qw(sum);
use autodie;

our @EXPORT = qw/ _prob_calc /;

# hashes containing Error probs
my %sanger_prob = (
    "\!"   => 1.000000000000000,
    "\“"   => 0.794328234700000,
    "\#"   => 0.630957344500000,
    "\$"   => 0.501187233600000,
    "\%"   => 0.398107170600000,
    "\&"   => 0.316227766000000,
    "\'"   => 0.251188643200000,
    "\("   => 0.199526231500000,
    "\)"   => 0.158489319200000,
    "\*"   => 0.125892541200000,
    "\+"   => 0.100000000000000,
    "\,"   => 0.079432823500000,
    "\-"   => 0.063095734400000,
    "\."   => 0.050118723400000,
    "\/"   => 0.039810717100000,
    "0"    => 0.031622776600000,
    "1"    => 0.025118864300000,
    "2"    => 0.019952623100000,
    "3"    => 0.015848931900000,
    "4"    => 0.012589254100000,
    "5"    => 0.010000000000000,
    "6"    => 0.007943282300000,
    "7"    => 0.006309573400000,
    "8"    => 0.005011872300000,
    "9"    => 0.003981071700000,
    "\:"   => 0.003162277700000,
    ";"    => 0.002511886400000,
    "\<"   => 0.001995262300000,
    "\="   => 0.001584893200000,
    "\>"   => 0.001258925400000,
    "\?"   => 0.001000000000000,
    "\@"   => 0.000794328200000,
    "A"    => 0.000630957300000,
    "B"    => 0.000501187200000,
    "C"    => 0.000398107200000,
    "D"    => 0.000316227800000,
    "E"    => 0.000251188600000,
    "F"    => 0.000199526200000,
    "G"    => 0.000158489300000,
    "H"    => 0.000125892500000,
    "I"    => 0.000100000000000,
    "J"    => 0.000079432800000,
    "K"    => 0.000063095700000,
    "L"    => 0.000050118700000,
    "M"    => 0.000039810700000,
    "N"    => 0.000031622800000,
    "O"    => 0.000025118900000,
    "P"    => 0.000019952600000,
    "Q"    => 0.000015848900000,
    "R"    => 0.000012589300000,
    "S"    => 0.000010000000000,
    "T"    => 0.000007943300000,
    "U"    => 0.000006309600000,
    "V"    => 0.000005011900000,
    "W"    => 0.000003981100000,
    "X"    => 0.000003162300000,
    "Y"    => 0.000002511900000,
    "Z"    => 0.000001995300000,
    "\["   => 0.000001584900000,
    "\\"   => 0.000001258900000,
    "\]"   => 0.000001000000000,
    "\^"   => 0.000000794300000,
    "\_"   => 0.000000631000000,
    "\`"   => 0.000000501200000,
    "a"    => 0.000000398100000,
    "b"    => 0.000000316200000,
    "c"    => 0.000000251200000,
    "d"    => 0.000000199500000,
    "e"    => 0.000000158500000,
    "f"    => 0.000000125900000,
    "g"    => 0.000000100000000,
    "h"    => 0.000000079400000,
    "i"    => 0.000000063100000,
    "j"    => 0.000000050100000,
    "k"    => 0.000000039800000,
    "l"    => 0.000000031600000,
    "m"    => 0.000000025100000,
    "n"    => 0.000000020000000,
    "o"    => 0.000000015800000,
    "p"    => 0.000000012600000,
    "q"    => 0.000000010000000,
    "r"    => 0.000000007900000,
    "s"    => 0.000000006300000,
    "t"    => 0.000000005000000,
    "u"    => 0.000000004000000,
    "v"    => 0.000000003200000,
    "w"    => 0.000000002500000,
    "x"    => 0.000000002000000,
    "y"    => 0.000000001600000,
    "z"    => 0.000000001300000,
    "\{"   => 0.000000001000000,
    "\|"   => 0.000000000800000,
    "\}"   => 0.000000000600000,
    "\~"   => 0.000000000500000
);

my %illumina_prob = (
    "@"  => 1.000000000000000,
    "A"  => 0.794328234700000,
    "B"  => 0.630957344500000,
    "C"  => 0.501187233600000,
    "D"  => 0.398107170600000,
    "E"  => 0.316227766000000,
    "F"  => 0.251188643200000,
    "G"  => 0.199526231500000,
    "H"  => 0.158489319200000,
    "I"  => 0.125892541200000,
    "J"  => 0.100000000000000,
    "K"  => 0.079432823500000,
    "L"  => 0.063095734400000,
    "M"  => 0.050118723400000,
    "N"  => 0.039810717100000,
    "O"  => 0.031622776600000,
    "P"  => 0.025118864300000,
    "Q"  => 0.019952623100000,
    "R"  => 0.015848931900000,
    "S"  => 0.012589254100000,
    "T"  => 0.010000000000000,
    "U"  => 0.007943282300000,
    "V"  => 0.006309573400000,
    "W"  => 0.005011872300000,
    "X"  => 0.003981071700000,
    "Y"  => 0.003162277700000,
    "Z"  => 0.002511886400000,
    "["  => 0.001995262300000,
    "\\" => 0.001584893200000,
    "]"  => 0.001258925400000,
    "\^" => 0.001000000000000,
    "\_" => 0.000794328200000,
    "\`" => 0.000630957300000,
    "a"  => 0.000501187200000,
    "b"  => 0.000398107200000,
    "c"  => 0.000316227800000,
    "d"  => 0.000251188600000,
    "e"  => 0.000199526200000,
    "f"  => 0.000158489300000,
    "g"  => 0.000125892500000,
    "h"  => 0.000100000000000,
    "i"  => 0.000079432800000,
    "j"  => 0.000063095700000,
    "k"  => 0.000050118700000,
    "l"  => 0.000039810700000,
    "m"  => 0.000031622800000,
    "n"  => 0.000025118900000,
    "o"  => 0.000019952600000,
    "p"  => 0.000015848900000,
    "q"  => 0.000012589300000,
    "r"  => 0.000010000000000,
    "s"  => 0.000007943300000,
    "t"  => 0.000006309600000,
    "u"  => 0.000005011900000,
    "v"  => 0.000003981100000,
    "w"  => 0.000003162300000,
    "x"  => 0.000002511900000,
    "y"  => 0.000001995300000,
    "z"  => 0.000001584900000,
    "{"  => 0.000001258900000,
    "|"  => 0.000001000000000,
    "}"  => 0.000000794300000,
    "~"  => 0.000000631000000
);

####################################

=head1 prob_calc

 Title   :  prob_calc
 Usage   :  $prob = prob_calc( $read, $lib, $cuttoff );
 Function:  dertmines error probability for each read
 Returns :  1 or 0 
 
=cut

sub _prob_calc{
    my $seq      = $_[0];
    my $lib      = $_[1];
    my $cuttoff  = $_[2];
    my $score_prob;
    my @prob_array;

        foreach my $i ( split //, $seq ) {
            if ( $lib eq "sanger" ) {
                push @prob_array, $sanger_prob{$i};
                $score_prob += $sanger_prob{$i};
            }
            elsif ( $lib eq "illumina" ) {
                push @prob_array, $illumina_prob{$i};
                $score_prob += $illumina_prob{$i};
            }
        }
        my $error_prob = ((sprintf "%.10f", sum(@prob_array))*100)/length $seq; 

       if ( $error_prob > $cuttoff ) {
       	    return 0;
        }
        else {
            return 1;
        }
    return 0;
}
####################################
####################################

1;


lib/fastQ_brew.pm
#
# module for fastQ_brew
#
# Copyright Damien O'Halloran
#
# You may distribute this module under the same terms as perl itself
# POD documentation - main docs before the code

=head1 NAME

fastQ_brew - a modern module for preprocessing of fastQ formatted files

=head1 DESCRIPTION

Provides methods for filtering and trimming reads by length and quality.

=head1 FEEDBACK

damienoh@gwu.edu

=head2 Mailing Lists

User feedback is an integral part of the evolution of this module. Send your comments and suggestions preferably to one of the mailing lists. Your participation is much appreciated.

=head2 Support

Please direct usage questions or support issues to:
<damienoh@gwu.edu>
Please include a thorough description of the problem with code and data examples if at all possible.

=head2 Reporting Bugs

Report bugs to the GitHub bug tracking system to help keep track of the bugs and their resolution.  Bug reports can be submitted via the GitHub page:

 https://github.com/dohalloran/fastQ_brew/issues

=head1 AUTHORS - Damien OHalloran

Email: damienoh@gwu.edu

=head1 APPENDIX

The rest of the documentation details each method

=cut

# Let the code begin...

package fastQ_brew;

use Moose;
with 'MooseX::Getopt', 'MooseX::Getopt::Usage::Role::Man';
use Modern::Perl;
use fastQ_brew_Utilities;
use fastQ_brew_Conversions;
use File::Copy qw(copy);
use Text::LevenshteinXS qw(distance);
use if ( $^O eq 'MSWin32' ), 'Win32::Console::ANSI';
use Term::ANSIColor qw(:constants);
use List::MoreUtils qw(any first_index);
use Sys::Hostname;
use autodie;

##################################
our $VERSION = '2.0';
##################################

=head2 fastQ_brew->new_with_options()

 Function: Populates the user data into $self 
 Returns : nothing returned
 Args    :
  --i, input file (required)
  --x, input file2 only to check demultiplexing 
  --plex, checks that two FASTQ files were demultiplexed correctly
  --o, output file (default is filtered.fq)
  --lib, library type  (default is sanger)       
  --qf, max. probability for a read to contain errors: suggested p<=50% (must be 1-100)
  --lf, filter by read length (suggested default =25)
  --trim_l, trim reads starting at left end
  --trim_r, trim reads starting at left end
  --truseq, removes truseq adapters from reads   
  --fasta, convert to fastA format 
  --dna_rna, convert reads to RNA
  --rev_comp, reverse complement reads
  --help, Print this help

=cut

##################################

has 'i'        	=> ( is => 'rw', isa => 'Str',  required => 1 );
has 'x'         => ( is => 'rw', isa => 'Str',  required => 0 );
has 'plex'      => ( is => 'rw', isa => 'Bool', default  => 0 );
has 'o'        	=> ( is => 'rw', isa => 'Str',  default  => "filtered.fq" );
has 'lib'      	=> ( is => 'rw', isa => 'Str',  default  => "sanger" );
has 'qf'       	=> ( is => 'rw', isa => 'Num',  default  => 0 );
has 'lf'       	=> ( is => 'rw', isa => 'Int',  default  => 0 );
has 'truseq'   	=> ( is => 'rw', isa => 'Bool', default  => 0 );
has 'trim_l'   	=> ( is => 'rw', isa => 'Int',  default  => 0 );
has 'trim_r'   	=> ( is => 'rw', isa => 'Int',  default  => 0 );
has 'fasta'    	=> ( is => 'rw', isa => 'Bool', default  => 0 );
has 'dna_rna'  	=> ( is => 'rw', isa => 'Bool', default  => 0 );
has 'rev_comp' 	=> ( is => 'rw', isa => 'Bool', default  => 0 );
has 'start'    	=> ( is => 'ro', isa => 'Int',  default  => time );
has help       	=> (
    is            	=> 'ro',
    isa           	=> 'Bool',
    default       	=> 0,
    documentation => qq{view driver script for input flag details}
);

##################################

=head2 truseq adapters

>TruSeq_Universal_Adapter
AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT
>TruSeq_Adapter_Index_1_6
GATCGGAAGAGCACACGTCTGAACTCCAGTCACATCACGATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_2
GATCGGAAGAGCACACGTCTGAACTCCAGTCACCGATGTATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_3
GATCGGAAGAGCACACGTCTGAACTCCAGTCACTTAGGCATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_4
GATCGGAAGAGCACACGTCTGAACTCCAGTCACTGACCAATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_5
GATCGGAAGAGCACACGTCTGAACTCCAGTCACACAGTGATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_6
GATCGGAAGAGCACACGTCTGAACTCCAGTCACGCCAATATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_7
GATCGGAAGAGCACACGTCTGAACTCCAGTCACCAGATCATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_8
GATCGGAAGAGCACACGTCTGAACTCCAGTCACACTTGAATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_9
GATCGGAAGAGCACACGTCTGAACTCCAGTCACGATCAGATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_10
GATCGGAAGAGCACACGTCTGAACTCCAGTCACTAGCTTATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_11
GATCGGAAGAGCACACGTCTGAACTCCAGTCACGGCTACATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_12
GATCGGAAGAGCACACGTCTGAACTCCAGTCACCTTGTAATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_13
GATCGGAAGAGCACACGTCTGAACTCCAGTCACAGTCAACAATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_14
GATCGGAAGAGCACACGTCTGAACTCCAGTCACAGTTCCGTATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_15
GATCGGAAGAGCACACGTCTGAACTCCAGTCACATGTCAGAATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_16
GATCGGAAGAGCACACGTCTGAACTCCAGTCACCCGTCCCGATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_18_7
GATCGGAAGAGCACACGTCTGAACTCCAGTCACGTCCGCACATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_19
GATCGGAAGAGCACACGTCTGAACTCCAGTCACGTGAAACGATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_20
GATCGGAAGAGCACACGTCTGAACTCCAGTCACGTGGCCTTATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_21
GATCGGAAGAGCACACGTCTGAACTCCAGTCACGTTTCGGAATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_22
GATCGGAAGAGCACACGTCTGAACTCCAGTCACCGTACGTAATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_23
GATCGGAAGAGCACACGTCTGAACTCCAGTCACGAGTGGATATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_25
GATCGGAAGAGCACACGTCTGAACTCCAGTCACACTGATATATCTCGTATGCCGTCTTCTGCTTG
>TruSeq_Adapter_Index_27
GATCGGAAGAGCACACGTCTGAACTCCAGTCACATTCCTTTATCTCGTATGCCGTCTTCTGCTTG

=cut

##################################

my %truseq_seqs = (
    0  => "AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT",
    1  => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACATCACGATCTCGTATGCCGTCTTCTGCTTG",
    2  => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACCGATGTATCTCGTATGCCGTCTTCTGCTTG",
    3  => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACTTAGGCATCTCGTATGCCGTCTTCTGCTTG",
    4  => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACTGACCAATCTCGTATGCCGTCTTCTGCTTG",
    5  => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACACAGTGATCTCGTATGCCGTCTTCTGCTTG",
    6  => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGCCAATATCTCGTATGCCGTCTTCTGCTTG",
    7  => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACCAGATCATCTCGTATGCCGTCTTCTGCTTG",
    8  => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACACTTGAATCTCGTATGCCGTCTTCTGCTTG",
    9  => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGATCAGATCTCGTATGCCGTCTTCTGCTTG",
    10 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACTAGCTTATCTCGTATGCCGTCTTCTGCTTG",
    11 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGGCTACATCTCGTATGCCGTCTTCTGCTTG",
    12 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACCTTGTAATCTCGTATGCCGTCTTCTGCTTG",
    13 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACAGTCAACAATCTCGTATGCCGTCTTCTGCTTG",
    14 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACAGTTCCGTATCTCGTATGCCGTCTTCTGCTTG",
    15 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACATGTCAGAATCTCGTATGCCGTCTTCTGCTTG",
    16 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACCCGTCCCGATCTCGTATGCCGTCTTCTGCTTG",
    17 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGTCCGCACATCTCGTATGCCGTCTTCTGCTTG",
    18 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGTGAAACGATCTCGTATGCCGTCTTCTGCTTG",
    19 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGTGGCCTTATCTCGTATGCCGTCTTCTGCTTG",
    20 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGTTTCGGAATCTCGTATGCCGTCTTCTGCTTG",
    21 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACCGTACGTAATCTCGTATGCCGTCTTCTGCTTG",
    22 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGAGTGGATATCTCGTATGCCGTCTTCTGCTTG",
    23 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACACTGATATATCTCGTATGCCGTCTTCTGCTTG",
    24 => "GATCGGAAGAGCACACGTCTGAACTCCAGTCACATTCCTTTATCTCGTATGCCGTCTTCTGCTTG"
);

##################################

=head2 run_fastQ_brew()

 Title   : run_fastQ_brew()
 Usage   : $self->run_fastQ_brew(%arg)
 Function: option to filter reads
 Returns : filtered fastQ file
 Args    : $self, %arg

=cut

##################################

sub run_fastQ_brew {
    my ( $self, %arg ) = @_;

    my ( $temp, @temp, $count, @number_matchesR, @number_matchesL );

    # log data on sys and infile
    my $inSize = -s $self->{i};
    print BOLD GREEN, "\nPerl Version: \t\t" . $^V;
    print "\nOperating system: \t" . $^O;
    print "\nHostname: \t\t" . hostname;
    print "\nfastQ_brew version: \t" . $VERSION;
    print "\nInput file size: \t" . $inSize . "KB", RESET;
    print BOLD CYAN, "\n\nFiltering file with fastQ_brew...", RESET;

    #check for file conversion calls or demultiplex check
    unless ( $self->{fasta} eq 0 ) {
        _convert_fasta( $self->{i} );
    }
    unless ( $self->{rev_comp} eq 0 ) {
        _rev_comp( $self->{i} );
    }
    unless ( $self->{dna_rna} eq 0 ) {
        _dna_rna( $self->{i} );
    }
    unless ( $self->{plex} eq 0 ) {
        _de_plex( $self->{i}, $self->{x} );
    }

    my $trimmed = "temp_";

    # new outfile resulting from filtering
    open my $fh_out, '>', $trimmed
      or die "Cannot open $trimmed: $!";

    open my $fh, '<', $self->{i}
      or die "Cannot open $self->{in_file}: $!";

    while (<$fh>) {
        $count++;
        chomp( $temp[0] = $_ );
        chomp( $temp[1] = <$fh> );
        chomp( $temp[2] = <$fh> );
        chomp( $temp[3] = <$fh> );

        my $key;

        ##  start by considering adapter removal
        if ( $self->{truseq} ne 0 ) {
            @number_matchesR = ();
            @number_matchesL = ();
            foreach $key ( sort { $a <=> $b } keys %truseq_seqs ) {
                my $hammingR = substr $temp[1], -length $truseq_seqs{$key},
                  length $truseq_seqs{$key};
                push @number_matchesR,
                  distance( $hammingR, $truseq_seqs{$key} );
                my $hammingL = substr $temp[1], 0, length $truseq_seqs{$key};
                push @number_matchesL,
                  distance( $hammingL, $truseq_seqs{$key} );
            }

            if (   any { $_ == 0 } @number_matchesR
                or any { $_ == 1 } @number_matchesR )
            {
                my $index = first_index { $_ < 2 } @number_matchesR;
                $temp[1]  = substr $temp[1], 0, -length $truseq_seqs{$index};
                $temp[3] = substr $temp[3], 0, -length $truseq_seqs{$index};
            }
            if (   any { $_ == 0 } @number_matchesL
                or any { $_ == 1 } @number_matchesL )
            {
                my $index = first_index { $_ < 2 } @number_matchesL;
                $temp[1] = substr $temp[1], length $truseq_seqs{$index},
                  length $temp[1];
                $temp[3] = substr $temp[3], length $truseq_seqs{$index},
                  length $temp[3];
            }
        }

        # if trim_l option not equal to N, then trim
        if ( $self->{trim_l} ne 0 ) {
            $temp[1] = substr $temp[1], $self->{trim_l}, length $temp[1];
            $temp[3] = substr $temp[3], $self->{trim_l},
              length $temp[3];

        }

        # if trim_r option not equal to N, then trim
        if ( $self->{trim_r} ne 0 ) {
            $temp[1]  = substr $temp[1],  0, -$self->{trim_r};
            $temp[3] = substr $temp[3], 0, -$self->{trim_r};

        }

        my $temp_prob;

        # get the read error probability score
        if ( $self->{qf} ne 0 ) {
            $temp_prob = _prob_calc( $temp[3], $self->{lib}, $self->{qf} );
        }
        else {
            $temp_prob = 1;
        }

        # conditionals for length
        if (
            (
                   ( length $temp[1] > $self->{lf} && $self->{lf} ne 0 )
                || ( $self->{lf} eq 0 )
            )
            && $temp_prob == 1
          )
        {
            print $fh_out "$temp[0]\n";
            print $fh_out "$temp[1]\n";
            print $fh_out "$temp[2]\n";
            print $fh_out "$temp[3]\n";
        }

    }
    $self->{original_reads} = $count;
    $self->{i}              = $trimmed;
    close $fh;
    close $fh_out;
    $self->_cleanup(%arg);
}

##################################

=head2 _cleanup()

 Title   : _cleanup()
 Usage   : _cleanup(%arg);
 Function: delete tmp files
 Returns : summary stats
 Args    : %arg

=cut

##################################

sub _cleanup {
    my ( $self, %arg ) = @_;
    my $count;
    my $clean_out = $self->{o};
    copy $self->{i}, $clean_out;
    my $size = -s $clean_out;

    open my $fh, '<', $clean_out
      or die "Cannot open $clean_out: $!";
    print BOLD YELLOW,
      "\n\nInput number of reads: \t" . $self->{original_reads};
    print "\nOutput file size: \t" . $size . "KB";
    unless ( $size == 0 ) {
        $count += tr/\n/\n/ while sysread( $fh, $_, 2**16 );
        my $outCount = $count / 4;
        print "\nOutput number of reads:\t" . $outCount;
        print "\nReads removed:\t\t"
          . ( ( $self->{original_reads} ) - ($outCount) );
        print "\nPercent removed:\t"
          . (sprintf "%.6f", (((( ( $self->{original_reads} ) - ($outCount) ))/$self->{original_reads})*100))."%";
    }
    my $duration = time - $self->{start};
    print "\nExecution time:\t\t" . $duration . "secs\n\n", RESET;

    # delete tmp files
    if ( -e "temp_" ) {
        unlink "temp_";
    }
    print BOLD RED, "fastQ_brew is now finished\n\n", RESET;
}

###################################

=head1 LICENSE AND COPYRIGHT

 Copyright (C) 2017 Damien M. O'Halloran
 GNU GENERAL PUBLIC LICENSE

=cut

1;
